lib/scholar/metrics/discounted_cumulative_gain.ex

```diff
@@ -0,0 +1,86 @@
+defmodule Scholar.Metrics.DiscountedCumulativeGain do
+  @moduledoc """
+  Discounted Cumulative Gain (DCG) is a measure of ranking quality.
+  It is based on the assumption that highly relevant documents appearing lower
+  in a search result list should be penalized as the graded relevance value is
+  reduced logarithmically proportional to the position of the result.
+  """
+
+  @doc """
+  Computes the DCG based on true relevancies (`y_true`) and predicted scores (`y_score`).
+  """
+
+  def compute(y_true, y_score, k \\ nil) do
+    # Ensure tensors are of the same shape
+    if Nx.shape(y_true) != Nx.shape(y_score) do
+      raise ArgumentError, "y_true and y_score tensors must have the same shape"
+    end
+
+    {adjusted_y_true, adjusted_y_score} = handle_ties(y_true, y_score)
+
+    sorted_indices = Nx.argsort(adjusted_y_score, axis: 0, direction: :desc)
+    sorted_y_true = Nx.take(adjusted_y_true, sorted_indices)
+
+    truncated_y_true = truncate_at_k(sorted_y_true, k)
+    dcg_value(truncated_y_true)
+  end
+  defp handle_ties(y_true, y_score) do
+    # Zip y_true and y_score together to work with pairs and convert to lists
+    zipped = y_true |> Nx.to_list() |> Enum.zip(Nx.to_list(y_score))
+
+    # Group items by their predicted scores and adjust groups if they contain ties
+    adjusted =
+      zipped
+      |> Enum.group_by(&elem(&1, 1))
+      |> Enum.flat_map(&adjust_group/1)
+
+    # Convert the lists back to tensors
+    {
+      Nx.tensor(Enum.map(adjusted, &elem(&1, 0))),
+      Nx.tensor(Enum.map(adjusted, &elem(&1, 1)))
+    }
+  end
```

Comment:
I believe you can avoid this conversion by using something like this:

```elixir
iex(6)> t = Nx.tensor([1, 2, 2, 3, 3, 3, 4])
#Nx.Tensor<
  s64[7]
  [1, 2, 2, 3, 3, 3, 4]
>
iex(7)> Nx.diff(t)
#Nx.Tensor<
  s64[6]
  [1, 0, 1, 0, 0, 1]
>
iex(8)> selector = Nx.pad(Nx.diff(t), 1, [{1, 0, 0}])
#Nx.Tensor<
  s64[7]
  [1, 1, 0, 1, 0, 0, 1]
>
iex(9)> Nx.select(selector, t, 0)
#Nx.Tensor<
  s64[7]
  [1, 2, 0, 3, 0, 0, 4]
>
```

where the select would set the proper placeholder and the effective could would be `Nx.sum(selector)`